SeptemberHX/joplin-plugin-enhancement
# Formatting bar: Add support for CodeMirror 6

# Summary

This pull request builds on #64 to allow the formatting bar to work with CodeMirror 6.

# Notes

- This pull request disables some functionality that was causing errors.
- The CodeMirror 6 formatting bar uses CodeMirror's built-in [tooltip functionality](https://codemirror.net/examples/tooltip/).
    - Currently, the formatting bar **is left-aligned** with the highlighted content in CodeMirror 6. This should be possible to fix with [the `Tooltip.offset` property](https://codemirror.net/docs/ref/#view.TooltipView.offset), but will require extra work to ensure that the formatting bar does not go offscreen.
- This pull request tries to avoid rebasing (as I've heard this can break in-progress code reviews). As such, it includes a few "revert changes to..." commits. On request, I can squash some of this pull request's commits into more meaningful commits.

# Testing

1. Enable the beta editor
2. Select text
3. Toggle bold formatting using the toolbar (on, then off)
4. Toggle italic formatting (on, then off)
5. Toggle code (on, then off)
6. Add a link to the selection
7. Select different text
8. Toggle all 7 colors
9. Repeat steps 2-8 for the legacy editor

Additional testing (done later):
1. Enable Quick Commands
2. Verify that the formatting bar can still be used in the CodeMirror 6-based editor
3. Switch to CodeMirror 5
4. Verify that Quick Commands can still be used

## Screen recording

https://github.com/SeptemberHX/joplin-plugin-enhancement/assets/46334387/81ccded2-b8b9-4c16-9d63-2eddc9068b11
